toolchain/check/handle_binding_pattern.cpp

```diff
@@ -216,23 +212,93 @@ static auto HandleAnyBindingPattern(Context& context, Parse::NodeId node_id,
+
+  auto [name_node, name_id] = context.node_stack().PopNameWithNodeId();
+  auto entity_name_id = context.entity_names().Add(
+      {.name_id = name_id,
+       .parent_scope_id = context.scope_stack().PeekNameScopeId()});
+
+  return HandleAnyBindingPattern(context, node_id, name_node, entity_name_id,
+                                 type_node, type_inst_id,
                                  Parse::NodeKind::LetBindingPattern);
 }
 
 auto HandleParseNode(Context& context, Parse::VarBindingPatternId node_id)
     -> bool {
-  return HandleAnyBindingPattern(context, node_id,
+  auto [type_node, type_inst_id] = context.node_stack().PopExprWithNodeId();
+
+  // `template` is incorrect here, but is diagnosed in parse.
+  context.node_stack()
+      .PopAndDiscardSoloNodeIdIf<Parse::NodeKind::TemplateBindingName>();
+
+  auto [name_node, name_id] = context.node_stack().PopNameWithNodeId();
+  auto entity_name_id = context.entity_names().Add(
+      {.name_id = name_id,
+       .parent_scope_id = context.scope_stack().PeekNameScopeId()});
+
+  return HandleAnyBindingPattern(context, node_id, name_node, entity_name_id,
+                                 type_node, type_inst_id,
                                  Parse::NodeKind::VarBindingPattern);
 }
 
 auto HandleParseNode(Context& context,
                      Parse::CompileTimeBindingPatternStartId node_id) -> bool {
-  // Make a scope to contain the `.Self` facet value for use in the type of the
-  // compile time binding. This is popped when handling the
-  // CompileTimeBindingPatternId.
-  context.scope_stack().PushForSameRegion();
+  auto is_template =
+      context.node_stack()
+          .PopAndDiscardSoloNodeIdIf<Parse::NodeKind::TemplateBindingName>();
+
+  auto name_id = context.node_stack().PeekNameId();
+  auto entity_name_id = SemIR::EntityNameId::None;
+
+  const DeclIntroducerState& introducer =
+      context.decl_introducer_state_stack().innermost();
+  if (introducer.kind == Lex::TokenKind::Let) {
+    // Disallow `let` outside of function and interface definitions.
+    // TODO: Find a less brittle way of doing this. A `scope_inst_id` of `None`
+    // can represent a block scope, but is also used for other kinds of scopes
+    // that aren't necessarily part of a function decl.
+    // We don't need to check if the scope is an interface here as this is
+    // already caught in the parse phase by the separated associated constant
+    // logic.
+    auto scope_inst_id = context.scope_stack().PeekInstId();
+    if (scope_inst_id.has_value()) {
+      auto scope_inst = context.insts().Get(scope_inst_id);
+      if (!scope_inst.Is<SemIR::FunctionDecl>()) {
+        context.TODO(
+            node_id,
+            "`let` compile time binding outside function or interface");
+        entity_name_id = context.entity_names().Add(
+            {.name_id = name_id,
+             .parent_scope_id = context.scope_stack().PeekNameScopeId()});
```

Comment:
Along this error path, we're eventually going to treat this binding as being a plain let binding, rather than a compile-time binding. I wonder if we should be returning early here, and skip creating the `.Self` in this case? That'd make our recovery more consistent with how we treat `:` bindings.

```suggestion
        context.TODO(
            node_id,
            "`let` compile time binding outside function or interface");
        context.scope_stack().PushForSameRegion();
        context.node_stack().Push(
            node_id,
            context.entity_names().Add(
                {.name_id = name_id,
                 .parent_scope_id = context.scope_stack().PeekNameScopeId()});
        return true;
```

(Maybe we could move the `PushForSameRegion` earlier rather than duplicating it.)

Reply:
That's a good plan, done.

Regarding the PushForSameRegion, it needs to come after the `context.scope_stack().AddCompileTimeBinding()`, so it can't be moved ahead of here.

Reply:
I figured we can actually just skip the PushForSameRegion when it's not needed. We are already looking to see which branch was taken when we later run the CompileTimeBindingPatternId handler.